#lamp-denv

We foster the openness, integrity, and reproducibility of scientific research.


##Associated publication

> **Development of four one-step real-time RT-LAMP assays for specific detection of dengue virus serotypes**
> 
> Benjamin Lopez-Jimena, Michaël Bekaert, Mohammed Bakheit, Sieghard Frischmann, Pranav Patel, Anavaj Sakuntabhai, Louis Lambrechts, Valerie D'Acremont, Manfred Weidmann.
>


##How to use this repository?

This repository host both the scripts and commands used by this study and the raw results generated at the time.


####Collecting genomes

The NCBI genomes were collected on 26th April 2014 and automatically aligned them with [GramAlign v3.0](http://bioinfo.unl.edu/gramalign.php)

```
./collect_genomevirus.pl -f denv1.fasta -d 1 -y 2004 -b 2014
./collect_genomevirus.pl -f denv2.fasta -d 2 -y 2004 -b 2014
./collect_genomevirus.pl -f denv3.fasta -d 3 -y 2004 -b 2014
./collect_genomevirus.pl -f denv4.fasta -d 4 -y 2004 -b 2014
```

where:

`-f` is the output file where the sequences will be saved.

`-d` is the dengue virus serotypes: 1, 2, 3 or 4.

`-y` or `-a` (after) is the lower limit for the year.

`-b` (before) is the upper limit for the year.


#### PCA and phylogeny

[R/adegenet](http://adegenet.r-forge.r-project.org/) was ran on the 4 strains to generate PCAs and phylogenies

```
Rscript denv.R --no-save --slave -q
```

#### Manual split into different subgroups/genotypes

Based of the PCA and phylogenetic analyses, the sequences were manually split into different subgroups/genotypes in order to maximise the region of sequence identity.

The TSV (Tabulation Separated Values) files were edited and two new columns (columns 5 and 6) with the genotype/group numbers you manually identified. The fifth column been a detailed subgoups (to be used with the loose parameter sets), the six column been the less fragmented (to be used with the standard parameter sets)

e.g.

```
	PC1	PC2	PC3	loose_group	standard_group
sequence_1	-14.75	-7.31	1.42	1	1
sequence_2	-14.16	9.83	1.20	1	1
sequence_3	-14.85	-7.83	1.10	2	1
sequence_4	-14.38	-6.75	1.06	2	1
...
```

#### Building the primer sets

Retrieve the subgroups/genotypes from the TSV file and run [LAVA-DNA](https://github.com/dylanstorey/lava-dna) on each genotype and each combination. DENV1, DENV2 and DENV3 used the `loose` parameter set from [LAVA-DNA](https://github.com/dylanstorey/lava-dna), while DENV4 used the more restrictive `standard` parameter set.

```
./class_sequences.pl -a denv1.results/denv1.fasta.align.fa -l denv1.results/denv1.pca.edited.tsv -c -e > denv1.results/denv1.loose.output
./class_sequences.pl -a denv2.results/denv2.fasta.align.fa -l denv2.results/denv2.pca.edited.tsv -c -e > denv2.results/denv2.loose.output
./class_sequences.pl -a denv3.results/denv3.fasta.align.fa -l denv3.results/denv3.pca.edited.tsv -c -e > denv3.results/denv3.loose.output
./class_sequences.pl -a denv4.results/denv4.fasta.align.fa -l denv4.results/denv4.pca.edited.tsv -c > denv4.results/denv4.output
```

where:

`-a` is the alignment generated by GramAlign (fasta-align format).

`-l` is the list of subgroups/genotypes (TSV file).

`-c` allows to test combination of genotypes and not only each genotypes.

`-e` allows for the `loose` parameter set rather than `standard` by [LAVA-DNA](https://github.com/dylanstorey/lava-dna).


#### Map each primer set

After evaluating the best minimal set of primers, each primer set was mapped

```
cd denv1.results
../map_lamp.pl -png -a denv1.fasta.align.fa \
 -p denv1.fasta.align.fa_1.primers \
 -p denv1.fasta.align.fa_11.primers \
 -p denv1.fasta.align.fa_19_20_8.primers \
 -p denv1.fasta.align.fa_13.primers \
 -p denv1.fasta.align.fa_10_16_17_18_20.primers \
 -p denv1.fasta.align.fa_22_9.primers \
 -p denv1.fasta.align.fa_20_21.primers \
 -p denv1.fasta.align.fa_3.primers \
 -p denv1.fasta.align.fa_4.primers \
 -p denv1.fasta.align.fa_12_17_8.primers \
 -p denv1.fasta.align.fa_18_7.primers \
 -p denv1.fasta.align.fa_5.primers \
 -p denv1.fasta.align.fa_6.primers \
 -p denv1.fasta.align.fa_15.primers \
 -p denv1.fasta.align.fa_14_2.primers \
 >denv1.png
cd ..
```

```
cd denv2.results
../map_lamp.pl -png -a denv2.fasta.align.fa \
 -p denv2.fasta.align.fa_17_18.primers \
 -p denv2.fasta.align.fa_15_16_20.primers \
 -p denv2.fasta.align.fa_11_19_9.primers \
 -p denv2.fasta.align.fa_10.primers \
 -p denv2.fasta.align.fa_13_16_20.primers \
 -p denv2.fasta.align.fa_14_19_7.primers \
 -p denv2.fasta.align.fa_12_14_18_19.primers \
 -p denv2.fasta.align.fa_3.primers \
 -p denv2.fasta.align.fa_4.primers \
 -p denv2.fasta.align.fa_2_9.primers \
 -p denv2.fasta.align.fa_5.primers \
 -p denv2.fasta.align.fa_17_6.primers \
 -p denv2.fasta.align.fa_19_8.primers \
 -p denv2.fasta.align.fa_1.primers \
 >denv2.png
cd ..
```

```
cd denv3.results
../map_lamp.pl -png -a denv3.fasta.align.fa \
 -p denv3.fasta.align.fa_1.primers \
 -p denv3.fasta.align.fa_8_9.primers \
 -p denv3.fasta.align.fa_10_11_12_13_14_15.primers \
 -p denv3.fasta.align.fa_6_7.primers \
 -p denv3.fasta.align.fa_14_15_5.primers \
 -p denv3.fasta.align.fa_2.primers \
 -p denv3.fasta.align.fa_12_4.primers \
 -p denv3.fasta.align.fa_11_3.primers \
 >denv3.png
cd ..
```

```
cd denv4.v2
../map_lamp.pl -png -a denv4.fasta.align.fa \
 -p denv4.fasta.align.fa_1_2.primers \
 -p denv4.fasta.align.fa_2_3.primers \
 -p denv4.fasta.align.fa_4.primers \
 >denv4.png
cd ..
```

##Issues

If you have any problems with or questions about the scripts, please contact us through a [GitHub issue](https://github.com/pseudogene/lamp-denv/issues).
Any issue related to the scientific results themselves must be done directly with the authors.


##Contributing

You are invited to contribute new features, fixes, or updates, large or small; we are always thrilled to receive pull requests, and do our best to process them as fast as we can.


##License and distribution


This code is distributed under the GNU GPL license v3. The documentation, raw data and work are licensed under a Creative Commons Attribution-ShareAlike 4.0 International License.​

