[![Build Status](https://travis-ci.org/pseudogene/lamp-denv.svg?branch=master)](https://travis-ci.org/pseudogene/lamp-denv)

#lamp-denv

We foster the openness, integrity, and reproducibility of scientific research.

Scripts and tools used to develop one-step real-time RT-LAMP assays for specific detection of dengue virus.

##Associated publication

> **Development of four one-step real-time RT-LAMP assays for specific detection of dengue virus serotypes**
> 
> Benjamin Lopez-Jimena, Michaël Bekaert, Mohammed Bakheit, Sieghard Frischmann, Pranav Patel, Anavaj Sakuntabhai, Louis Lambrechts, Valerie D'Acremont, Manfred Weidmann.
>


##How to use this repository?

This repository host both the scripts and tools used by this study and the raw results generated at the time. Feel free to adapt the scripts and tools, but remember to cite their authors!

To look at our scripts and raw results, **browse** through this repository. If you want to reproduce our results you will need to **clone** this repository, build the docker, and the run all the scripts. If you want to use our data for our own research, **fork** this repository and **cite** the authors.


##Prepare a docker

All required files and tools run in a self-contained [docker](https://www.docker.com/) image.

####Clone the repository

```
git clone https://github.com/pseudogene/lamp-denv.git
cd lamp-denv
```

####Create a docker

```
docker build --rm=true --file=Dockerfile -t denv-docker .
```

####Start the docker

To import and export the results of your analyse you need to link a folder to the docker. It this example your data will be store in `results` (current filesystem) which will be seem as been `/mnt` from within the docker by using `-v <USERFOLDER>:/mnt`.

```
mkdir ~/results
docker run -i -t -v ~/results:/mnt --name dengue denv-docker /bin/bash
```


####Run a new analysis

**1** - Collect NCBI genomes and automatically aligned them with [GramAlign v3.0](http://bioinfo.unl.edu/gramalign.php)

```
./collect_genomevirus.pl -f denv1.fasta -d 1 -y 2004 -b 2014
./collect_genomevirus.pl -f denv2.fasta -d 2 -y 2004 -b 2014
./collect_genomevirus.pl -f denv3.fasta -d 3 -y 2004 -b 2014
./collect_genomevirus.pl -f denv4.fasta -d 4 -y 2004 -b 2014
```

where:

`-f` is the output file where the sequences will be saved.

`-d` is the dengue virus serotypes: 1, 2, 3 or 4.

`-y` or `-a` (after) is the lower limit for the year.

`-b` (before) is the upper limit for the year.


**2** - Run [R/adegenet](http://adegenet.r-forge.r-project.org/) on the 4 strains to generate PCA and phylogeny

```
Rscript denv.R --no-save --slave -q
cp -r denv?.* /mnt
```

**3** - In `/mnt` (from within the docker) or `results` (from outside) you now have all the results files including PCA, phylogeny and TSV files.

Edit the TSV (Tabulation Separated Values) files by adding two new columns (columns 5 and 6) with the genotype/group numbers you manually identified. The fifth column been a detailed subgoups (to be used with the loose parameter sets), the six column been the less fragmented (to be used with the standard parameter sets)

e.g.

```
	PC1	PC2	PC3
sequence_1	-14.75	-7.31	1.42
sequence_2	-14.16	9.83	1.00
```

becomes:

```
	PC1	PC2	PC3	group	group
sequence_1	-14.75	-7.31	1.42	1	1
sequence_2	-14.16	9.83	1.00	2	2
```

**4 **- Import your data. Copy the modified TSV in `/mnt` (from within the docker) or `results` (from outside)

```
dos2unix -c mac *.tsv
dos2unix *.tsv
cp *.tsv /dengue
```

**5** - Retrieve the subgroups/genotypes from the TSV file and run [LAVA-DNA](https://github.com/dylanstorey/lava-dna) on each genotype and each combination.

```
./class_sequences.pl -a denv1.fasta.align.fa -l denv1.pca.edited.tsv -c > denv1.log
./class_sequences.pl -a denv2.fasta.align.fa -l denv2.pca.edited.tsv -c -e > denv2.loose.log
./class_sequences.pl -a denv3.fasta.align.fa -l denv3.pca.edited.tsv -c > denv3.log
./class_sequences.pl -a denv4.fasta.align.fa -l denv4.pca.edited.tsv -c -e > denv4.loose.log
```

where:

`-a` is the alignment generated by GramAlign (fasta-align format).

`-l` is the list of subgroups/genotypes (TSV file).

`-c` allows to test combination of genotypes and not only each genotypes.

`-e` allows for the `loose` parameter set rather than `standard` by [LAVA-DNA](https://github.com/dylanstorey/lava-dna).


**6** - Then evaluate the best minimal set of primer and map each primer set

```
cd denv1
../map_lamp.pl -svg -a denv1.fasta.align.fa \
 -p denv1.fasta.align.fa_1.primers \
 -p denv1.fasta.align.fa_11.primers \
 -p denv1.fasta.align.fa_19_20_8.primers \
 -p denv1.fasta.align.fa_13.primers \
 > denv1.svg
cd ..

```

where:

`-svg` or `-png` will generate a SVG or PNG image.

`-a` is the alignment generated by GramAlign (fasta-align format).

`-p` are lists of primer sets generated by `./class_sequences.pl`.


##Issues

If you have any problems with or questions about the scripts, please contact us through a [GitHub issue](https://github.com/pseudogene/lamp-denv/issues).
Any issue related to the scientific results themselves must be done directly with the authors.


##Contributing

You are invited to contribute new features, fixes, or updates, large or small; we are always thrilled to receive pull requests, and do our best to process them as fast as we can.


##License and distribution


This code is distributed under the GNU GPL license v3. The documentation, raw data and work are licensed under a Creative Commons Attribution-ShareAlike 4.0 International License.​

